regression/rt_gk_sheath_1x2v_p1.c

```diff
@@ -332,28 +402,18 @@ bmag_func(double t, const double* GKYL_RESTRICT zc, double* GKYL_RESTRICT fout,
   fout[0] = B0;
 }
 
-void
-calc_integrated_diagnostics(struct gkyl_tm_trigger* iot, gkyl_gyrokinetic_app* app, double t_curr, bool force_calc)
-{
-  if (gkyl_tm_trigger_check_and_bump(iot, t_curr) || force_calc) {
-    gkyl_gyrokinetic_app_calc_field_energy(app, t_curr);
-    gkyl_gyrokinetic_app_calc_integrated_mom(app, t_curr);
-  }
-}
-
 void
 write_data(struct gkyl_tm_trigger* iot, gkyl_gyrokinetic_app* app, double t_curr, bool force_write)
 {
-  bool trig_now = gkyl_tm_trigger_check_and_bump(iot, t_curr);
-  if (trig_now || force_write) {
-    int frame = (!trig_now) && force_write? iot->curr : iot->curr-1;
-
-    gkyl_gyrokinetic_app_write(app, t_curr, frame);
+  if (gkyl_tm_trigger_check_and_bump(iot, t_curr)) {
+    int frame = iot->curr - 1;
+    if (force_write) {
+      frame = iot->curr;
+    }
 
-    gkyl_gyrokinetic_app_calc_field_energy(app, t_curr);
-    gkyl_gyrokinetic_app_write_field_energy(app);
+    gkyl_gyrokinetic_app_write(app, t_curr, iot->curr - 1);
 
-    gkyl_gyrokinetic_app_calc_integrated_mom(app, t_curr);
+    gkyl_gyrokinetic_app_calc_integrated_mom(app, iot->curr - 1);
     gkyl_gyrokinetic_app_write_integrated_mom(app);
   }
```

Comment:
the logic here was carefully chosen so that it produced the intended behavior (for both grid and integrated diagnostics) in:
1. Well behaved single job simulations
2. Restarted simulations
3. Crashing simulations
Did you check that you are reproducing the behavior we had before for these 3 scenarios?

Also, why did you remove the field_energy integrated diag?

Reply:
The boilerplate that has been introduced into the write method here was the boilerplate that you yourself wrote, and which was subsequently standardized across the entire G0 system. Are you saying that it was subsequently discovered that this code was inadequate in some regard (either for restarts or crashes)? If so, why wasn't it modified globally rather than just in GK?

If this is the case, then we should establish precisely what was wrong and why this change is necessary, and then make this change _globally_. And we should figure out why it wasn't caught before.

As for why the field energy integrated diagnostics were removed: there is no principled reason, and we can add these to the Lua app driver (in which case it makes sense to add them to the C tests too), but then we would need to decide whether they would be calculated and written for all tests, or whether we want some kind of Lua interface that determines when and if they are calculated and written, and a design for this needs to be established.

Reply:
But the logic in the GK reg tests here seems to have changed slightly compared to what I wrote (see for example rt_gk_sheath_2x2v_p1 in gk-g0-app), and my memory of it is that, after some trial and error, that exact logic is what was needed to produce the intended behavior in the 3 cases mentioned above for both grid and integrated diagnostics. Maybe I'm wrong, or it's possible that the new logic used here is equivalent, but I would just like to know if it is and/or if it behaves the same way.

Field energy diag: I would put it in every reg test. It's a useful tool (because it's more sensitive than the integrated moments). I thought that was the case already, and if it wasn't I'm sorry I didn't update that.